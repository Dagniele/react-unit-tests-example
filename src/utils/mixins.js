export function transition(what = "all", time = 0.2, way = "ease-in-out") {
  return `transition: ${what} ${time}s ${way};`;
}

export function noFocus() {
  return `
    &:focus {
      outline: none;
    }
  `;
}

const SHADOWS = {
  0: "none",
  2: "0 2px 2px 0 rgba(0, 0, 0, 0.14), 0 3px 1px -2px rgba(0, 0, 0, 0.2), 0 1px 5px 0 rgba(0, 0, 0, 0.12)",
  3: "0 3px 4px 0 rgba(0, 0, 0, 0.14), 0 3px 3px -2px rgba(0, 0, 0, 0.2), 0 1px 8px 0 rgba(0, 0, 0, 0.12)",
  4: "0 4px 5px 0 rgba(0, 0, 0, 0.14), 0 1px 10px 0 rgba(0, 0, 0, 0.12), 0 2px 4px -1px rgba(0, 0, 0, 0.2)",
  6: "0 6px 10px 0 rgba(0, 0, 0, 0.14), 0 1px 18px 0 rgba(0, 0, 0, 0.12), 0 3px 5px -1px rgba(0, 0, 0, 0.2)",
  8: "0 8px 10px 1px rgba(0, 0, 0, 0.14), 0 3px 14px 2px rgba(0, 0, 0, 0.12), 0 5px 5px -3px rgba(0, 0, 0, 0.2)",
  16: "0 16px 24px 2px rgba(0, 0, 0, 0.14), 0  6px 30px 5px rgba(0, 0, 0, 0.12), 0  8px 10px -5px rgba(0, 0, 0, 0.2)",
  24: "0  9px 46px  8px rgba(0, 0, 0, 0.14), 0 11px 15px -7px rgba(0, 0, 0, 0.12), 0 24px 38px  3px rgba(0, 0, 0, 0.2)"
};

export function makeShadow(dp) {
  if (!SHADOWS[dp]) throw new Error(`dp ${dp} is an invalid shadow depth`);

  return `box-shadow: ${SHADOWS[dp]};`;
}
