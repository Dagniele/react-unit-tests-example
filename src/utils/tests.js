import React from 'react';
import 'jest-styled-components';
import Enzyme, { mount as enzymeMount, shallow as enzymeShallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import { ThemeProvider } from 'styled-components';
import defaultTheme from '../theme';

Enzyme.configure({ adapter: new Adapter() });

/**
 * This function uses react-test-renderer which is less powerful of enzyme but way easier and
 * generates smaller snapshots. Most of the times it will be enough to use this one for testing
 */
export function rendererWithTheme(children, options = {}) {
  const tree = renderer.create(<ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>);

  if (!options.pure) return tree.toJSON();

  return tree;
}

/**
 * Why use enzyme utils from this file instead of importing straight from enzyme library?
 * Because enzyme needs the react adapter, so instead of including it in every test we
 * include it in this file
 */

/**
 * The next functions uses enzyme which gives you access to every mounted sub component,
 * useful in case you need to test something specific or you need to change the state
 * from outside, without simulating clicks or anything else. Mainly this should be
 * used when you don't want to just use the snapshot for testing, check "Button"
 * tests for few examples
 */
export function mountWithTheme(children) {
  const tree = enzymeMount(<ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>);
  return tree.childAt(0).childAt(0);
}

/**
 * This is like the previous one but it has no theme, usefult if we have to test user
 * interactions behaviors
 */
export function mount(children) {
  return enzymeMount(children);
}

/**
 * This is like the previous one but it has no theme, usefult if we have to test user
 * interactions behaviors
 */
export function shallow(children) {
  return enzymeShallow(children);
}
