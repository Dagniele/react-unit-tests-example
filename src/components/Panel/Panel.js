import React from "react";
import styled from "styled-components";
import { makeShadow } from "../../utils/mixins";

const StyledPanel = styled.div`
  ${props => makeShadow(props.dp)};
  padding: 10px;
  margin-bottom: 8px;
`;

const Panel = props => {
  const { dp = 2 } = props;
  return <StyledPanel dp={dp}>{props.children}</StyledPanel>;
};

export default Panel;
