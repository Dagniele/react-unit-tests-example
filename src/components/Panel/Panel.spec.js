import React from "react";
import { rendererWithTheme } from "../../utils/tests";
import Panel from "./Panel";

describe("Panel", () => {
  it("should match the snapshot", () => {
    const tree = rendererWithTheme(<Panel>My content</Panel>);
    expect(tree).toMatchSnapshot();
  });
});
