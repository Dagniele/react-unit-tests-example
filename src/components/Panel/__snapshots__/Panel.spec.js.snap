// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel should match the snapshot 1`] = `
.c0 {
  box-shadow: 0 2px 2px 0 rgba(0,0,0,0.14),0 3px 1px -2px rgba(0,0,0,0.2),0 1px 5px 0 rgba(0,0,0,0.12);
  padding: 10px;
  margin-bottom: 8px;
}

<div
  className="c0"
>
  My content
</div>
`;
